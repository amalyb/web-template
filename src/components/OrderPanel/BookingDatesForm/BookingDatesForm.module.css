@import '../../../styles/propertySets.css';

.root {
  display: flex;
  flex-direction: column;
}

.bookingDates {
  flex-shrink: 0;

  /* Ensure that calendar dropdown gets some stacking context relative to other form items below */
  z-index: 1;
}

.priceBreakdownContainer {
  padding: 0 24px;
  margin-top: 24px;

  @media (--viewportMedium) {
    padding: 0;
    margin-top: 40px;
  }
}

.error {
  color: var(--failColor);
  margin: 0 24px;
  display: inline-block;
}

.sideBarError {
  @apply --marketplaceH4FontStyles;
  color: var(--failColor);
  margin: 0 24px 12px 24px;

  @media (--viewportMedium) {
    margin: 0 0 12px 0;
  }
}

.finePrint {
  @apply --marketplaceTinyFontStyles;
  text-align: center;
}

.submitButton {
  padding: 0 24px;
  margin-top: 24px;

  @media (--viewportMedium) {
    padding: 0;
    margin-top: 72px;
  }
}

.arrowIcon {
  stroke: var(--matterColorLight);
  fill: var(--matterColorLight);
}
