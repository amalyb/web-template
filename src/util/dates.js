import moment from 'moment-timezone/builds/moment-timezone-with-data-10-year-range.min';

/**
 * Input names for the DateRangePicker from react-dates.
 */
export const START_DATE = 'startDate';
export const END_DATE = 'endDate';

/**
 * Check that the given parameter is a Date object.
 *
 * @param {Date} object that should be a Date.
 *
 * @returns {boolean} true if given parameter is a Date object.
 */
export const isDate = d =>
  d && Object.prototype.toString.call(d) === '[object Date]' && !Number.isNaN(d.getTime());

/**
 * Check if the given parameters represent the same Date value (timestamps are compared)
 *
 * @param {Date} first param that should be a Date and it should have same timestamp as second param.
 * @param {Date} second param that should be a Date and it should have same timestamp as second param.
 *
 * @returns {boolean} true if given parameters have the same timestamp.
 */
export const isSameDate = (a, b) => a && isDate(a) && b && isDate(b) && a.getTime() === b.getTime();

/**
 * Compare is dateA is after dateB
 *
 * @param {Date} dateA date instance
 * @param {Date} dateB date instance
 *
 * @returns {Date} true if dateA is after dateB
 */
export const isAfterDate = (dateA, dateB) => {
  return moment(storedAt).isAfter(moment(dateB));
};

////////////////////////////////////////////////////////////////////
// Manipulate time: time-of-day between different time zones etc. //
////////////////////////////////////////////////////////////////////

/**
 * Returns a new date, which indicates the same time of day in a given time zone
 * as given date is in local time zone
 *
 * @param {Date} date
 * @param {String} timeZone
 *
 * @returns {Date} date in given time zone
 */
export const timeOfDayFromLocalToTimeZone = (date, timeZone) => {
  return moment.tz(moment(date).format('YYYY-MM-DD HH:mm:ss'), timeZone).toDate();
};

/**
 * Returns a new date, which indicates the same time of day in a local time zone
 * as given date is in specified time zone
 *
 * @param {Date} date
 * @param {String} timeZone
 *
 * @returns {Date} date in given time zone
 */
export const timeOfDayFromTimeZoneToLocal = (date, timeZone) => {
  return moment(
    moment(date)
      .tz(timeZone)
      .format('YYYY-MM-DD HH:mm:ss')
  ).toDate();
};

/**
 * Get start of time unit (e.g. start of day)
 *
 * @param {Date} date date instance to be converted
 * @param {String} unit time-unit (e.g. "day")
 * @param {String} timeZone time zone id
 *
 * @returns {Date} date object converted to the start of given unit
 */
export const getStartOf = (date, unit, timeZone) => {
  const m = timeZone
    ? moment(date)
        .clone()
        .tz(timeZone)
    : moment(date).clone();

  return m.startOf(unit).toDate();
};

/**
 * Adds time-units to the date
 *
 * @param {Date} date date to be manipulated
 * @param {int} offset offset of time-units (e.g. "3" days)
 * @param {String} unit time-unit (e.g. "days")
 * @param {String} timeZone time zone name
 *
 * @returns {Date} date with given offset added
 */
export const addTime = (date, offset, unit, timeZone) => {
  const m = timeZone
    ? moment(date)
        .clone()
        .tz(timeZone)
    : moment(date).clone();
  return m.add(offset, unit).toDate();
};

/**
 * Subtract time-units from the date
 *
 * @param {Date} date date to be manipulated
 * @param {int} offset offset of time-units (e.g. "3" days)
 * @param {String} unit time-unit (e.g. "days")
 * @param {String} timeZone time zone name
 *
 * @returns {Date} date with given offset subtracted
 */
export const subtractTime = (date, offset, unit, timeZone) => {
  const m = timeZone
    ? moment(date)
        .clone()
        .tz(timeZone)
    : moment(date).clone();
  return m.subtract(offset, unit).toDate();
};

///////////////
// Durations //
///////////////

/**
 * Calculate the number of days between the given dates
 *
 * @param {Date} startDate start of the time period
 * @param {Date} endDate end of the time period. NOTE: with daily
 * bookings, it is expected that this date is the exclusive end date,
 * i.e. the last day of the booking is the previous date of this end
 * date.
 *
 * @throws Will throw if the end date is before the start date
 * @returns {Number} number of days between the given dates
 */
export const daysBetween = (startDate, endDate) => {
  const days = moment(endDate).diff(startDate, 'days');
  if (days < 0) {
    throw new Error('End date cannot be before start date');
  }
  return days;
};

/**
 * Calculate the number of minutes between the given dates
 *
 * @param {Date} startDate start of the time period
 * @param {Date} endDate end of the time period.
 *
 * @throws Will throw if the end date is before the start date
 * @returns {Number} number of minutes between the given Date objects
 */
export const minutesBetween = (startDate, endDate) => {
  const minutes = moment(endDate).diff(startDate, 'minutes');
  if (minutes < 0) {
    throw new Error('End Date cannot be before start Date');
  }
  return minutes;
};

/**
 * Calculate the difference between the given dates
 *
 * @param {Date} startDate start of the time period
 * @param {Date} endDate end of the time period.
 *
 * @returns {Number} time difference between the given Date objects using given unit
 */
export const diffInTime = (startDate, endDate, unit, useFloat = false) => {
  return startDate.diff(endDate, unit, useFloat);
};

////////////////////////////
// Parsing and formatting //
////////////////////////////

/**
 * Format the given date
 *
 * @param {Object} intl Intl object from react-intl
 * @param {String} todayString translation for the current day
 * @param {Date} d Date to be formatted
 *
 * @returns {String} formatted date
 */
export const formatDate = (intl, todayString, d) => {
  const paramsValid = intl && d instanceof Date && typeof todayString === 'string';
  if (!paramsValid) {
    throw new Error(`Invalid params for formatDate: (${intl}, ${todayString}, ${d})`);
  }

  // By default we can use moment() directly but in tests we need to use a specific dates.
  // fakeIntl used in tests contains now() function wich returns predefined date
  const now = intl.now ? moment(intl.now()) : moment();
  const formattedTime = intl.formatTime(d);
  let formattedDate;

  if (now.isSame(d, 'day')) {
    // e.g. "Today, 9:10pm"
    formattedDate = todayString;
  } else if (now.isSame(d, 'week')) {
    // e.g. "Wed, 8:00pm"
    formattedDate = intl.formatDate(d, {
      weekday: 'short',
    });
  } else if (now.isSame(d, 'year')) {
    // e.g. "Aug 22, 7:40pm"
    formattedDate = intl.formatDate(d, {
      month: 'short',
      day: 'numeric',
    });
  } else {
    // e.g. "Jul 17 2016, 6:02pm"
    const date = intl.formatDate(d, {
      month: 'short',
      day: 'numeric',
    });
    const year = intl.formatDate(d, {
      year: 'numeric',
    });
    formattedDate = `${date} ${year}`;
  }

  return `${formattedDate}, ${formattedTime}`;
};

/**
 * Converts string given in ISO8601 format to date object.
 * This is used e.g. when when dates are parsed form urlParams
 *
 * @param {String} dateString in 'YYYY-MM-DD'format
 *
 * @returns {Date} parsed date object
 */
export const parseDateFromISO8601 = dateString => {
  return moment(dateString, 'YYYY-MM-DD').toDate();
};

/**
 * Converts date to string ISO8601 format ('YYYY-MM-DD').
 * This string is used e.g. in urlParam.
 *
 * @param {Date} date
 *
 * @returns {String} string in 'YYYY-MM-DD'format
 */
export const stringifyDateToISO8601 = date => {
  return moment(date).format('YYYY-MM-DD');
};

/**
 * Formats string ('YYYY-MM-DD') to UTC format ('0000-00-00T00:00:00.000Z').
 * This is used in search query.
 *
 * @param {String} string in 'YYYY-MM-DD'format
 *
 * @returns {String} string in '0000-00-00T00:00:00.000Z' format
 */
export const formatDateStringToUTC = dateString => {
  return moment.utc(dateString).toDate();
};

/**
 * Formats date to into multiple different ways:
 * - date "Mar 24"
 * - time "8:07 PM"
 * - dateAndTime: "Mar 24, 8:07 PM"
 *
 * @param {Object} intl Intl object from react-intl
 * @param {Date} date to be formatted
 *
 * @returns {Object} "{ date, time, dateAndTime }"
 */
export const formatDateToText = (intl, date) => {
  return {
    date: intl.formatDate(date, {
      month: 'short',
      day: 'numeric',
    }),
    time: intl.formatDate(date, {
      hour: 'numeric',
      minute: 'numeric',
    }),
    dateAndTime: intl.formatTime(date, {
      month: 'short',
      day: 'numeric',
    }),
  };
};

//////////
// Misc //
//////////

/**
 * Format the given date to month id/string
 *
 * @param {Date} date to be formatted
 *
 * @returns {String} formatted month string
 */
export const monthIdString = date => moment(date).format('YYYY-MM');

/**
 * Format the given date to UTC month id/string
 *
 * @param {Date} date to be formatted
 *
 * @returns {String} formatted month string
 */
export const monthIdStringInUTC = date =>
  moment(date)
    .utc()
    .format('YYYY-MM');

/**
 * Formats string ('YYYY-MM-DD') to UTC format ('0000-00-00T00:00:00.000Z') and adds one day.
 * This is used as end date of the search query.
 * One day must be added because end of the availability is exclusive in API.
 *
 * @param {String} string in 'YYYY-MM-DD'format
 *
 * @returns {String} string in '0000-00-00T00:00:00.000Z' format
 */

export const getExclusiveEndDate = dateString => {
  return moment
    .utc(dateString)
    .add(1, 'days')
    .startOf('day')
    .toDate();
};
