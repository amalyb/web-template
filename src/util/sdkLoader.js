import * as importedSdk from 'sharetribe-flex-sdk';

let exportSdk;

const isServer = () => typeof window === 'undefined';

if (isServer()) {
  // Use eval to skip webpack from bundling SDK in Node
  // eslint-disable-next-line no-eval
  exportSdk = eval('require')('sharetribe-flex-sdk');
} else {
  exportSdk = importedSdk;
}

const { createInstance, types, transit, util } = exportSdk;

// create image variant from variant name, desired width and aspectRatio
const createImageVariantConfig = (name, width, aspectRatio) => {
  //const aspectRatio = aspectHeight / aspectWidth;
  return {
    [`imageVariant.${name}`]: util.objectQueryString({
      w: width,
      h: aspectRatio * width,
      fit: 'crop',
    }),
  };
};

export { createInstance, types, transit, util, createImageVariantConfig };
