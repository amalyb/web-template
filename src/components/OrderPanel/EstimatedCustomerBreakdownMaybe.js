/**
 * Booking breakdown estimation
 *
 * Transactions have payment information that can be shown with the
 * OrderBreakdown component. However, when selecting booking
 * details, there is no transaction object present and we have to
 * estimate the breakdown of the transaction without data from the
 * API.
 *
 * If the payment process of a customized marketplace is something
 * else than simply daily or nightly bookings, the estimation will
 * most likely need some changes.
 *
 * To customize the estimation, first change the BookingDatesForm to
 * collect all booking information from the user (in addition to the
 * default date pickers), and provide that data to the
 * EstimatedBreakdownMaybe components. You can then make customization
 * within this file to create a fake transaction object that
 * calculates the breakdown information correctly according to the
 * process.
 *
 * In the future, the optimal scenario would be to use the same
 * transactions.initiateSpeculative API endpoint as the CheckoutPage
 * is using to get the breakdown information from the API, but
 * currently the API doesn't support that for logged out users, and we
 * are forced to estimate the information here.
 */
import React, { useEffect } from 'react';
import Decimal from 'decimal.js';

import { types as sdkTypes, transit } from '../../util/sdkLoader';
import { FormattedMessage, useIntl } from '../../util/reactIntl';
import { LINE_ITEM_DAY, LINE_ITEM_NIGHT, LISTING_UNIT_TYPES } from '../../util/types';
import { unitDivisor, convertMoneyToNumber, convertUnitToSubUnit, formatMoney } from '../../util/currency';
import { getProcess, TX_TRANSITION_ACTOR_CUSTOMER } from '../../transactions/transaction';

import { OrderBreakdown } from '../../components';

import css from './OrderPanel.module.css';

import { differenceInCalendarDays } from 'date-fns'; // If not already imported


const { Money, UUID } = sdkTypes;

// Type handlers for defensive deserialization
const typeHandlers = [
  {
    type: sdkTypes.BigDecimal,
    customType: Decimal,
    writer: v => new sdkTypes.BigDecimal(v.toString()),
    reader: v => new Decimal(v.value),
  },
];

// Defensive guard to ensure lineItems is always an array
const fromTransit = str => transit.read(str, { typeHandlers });

const ensureArray = (li) => {
  if (Array.isArray(li)) return li;
  if (typeof li === 'string') return fromTransit(li);
  // if it's already a plain object (rare), wrap or normalize as needed
  return Array.isArray(li?.lineItems) ? li.lineItems : [];
};

// Add DOM-based logging
const domLog = (label, data) => {
  // Commenting out DOM logging to prevent browser display
  // if (typeof document === 'undefined') return;
  
  // const logDiv = document.getElementById('debug-log') || (() => {
  //   const div = document.createElement('div');
  //   div.id = 'debug-log';
  //   div.style.cssText = 'position: fixed; bottom: 0; right: 0; max-height: 200px; overflow: auto; background: white; border: 1px solid black; padding: 10px; z-index: 9999; font-family: monospace;';
  //   document.body.appendChild(div);
  //   return div;
  // })();

  // const entry = document.createElement('div');
  // entry.style.borderBottom = '1px solid #eee';
  // entry.innerHTML = `
  //   <strong>${new Date().toISOString()} - ${label}</strong><br/>
  //   <pre>${JSON.stringify(data, (key, value) => {
  //     if (value instanceof Money) {
  //       return `Money(${value.amount}, ${value.currency})`;
  //     }
  //     return value;
  //   }, 2)}</pre>
  // `;
  // logDiv.insertBefore(entry, logDiv.firstChild);
};

const debugLog = (label, data) => {
  try {
    window._debug = window._debug || [];
    const logEntry = {
      timestamp: new Date().toISOString(),
      label,
      data: JSON.stringify(data, (key, value) => {
        if (value instanceof Money) {
          return `Money(${value.amount}, ${value.currency})`;
        }
        return value;
      }, 2)
    };
    window._debug.push(logEntry);
    window.console.log(`[DEBUG] ${label}:`, data);
    // domLog(label, data); // Commenting out DOM logging
  } catch (e) {
    window.console.error('Debug logging failed:', e);
    // domLog('Error', e.toString()); // Commenting out DOM logging
  }
};

// Add global error handler for Money type issues
if (typeof window !== 'undefined') {
  window.onerror = function(msg, url, lineNo, columnNo, error) {
    if (msg.includes('Money type')) {
      debugLog('Money Type Error', {
        message: msg,
        url,
        lineNo,
        columnNo,
        error: error?.toString(),
        stack: error?.stack
      });
    }
    return false;
  };
}

// Add SDK validation
const validateSDK = () => {
  debugLog('Validating SDK', {
    sdkTypes,
    Money: sdkTypes.Money,
    isMoneyDefined: typeof Money === 'function',
    MoneyPrototype: Money?.prototype
  });

  if (!sdkTypes || typeof Money !== 'function') {
    debugLog('SDK Error', 'SDK or Money constructor not properly initialized');
    return false;
  }
  return true;
};

// Update Money validation
const validateMoneyObject = (obj, context) => {
  if (!validateSDK()) {
    return false;
  }

  debugLog(`Validating Money (${context})`, {
    value: obj,
    type: obj?.constructor?.name,
    isMoneyInstance: obj instanceof Money,
    hasAmount: obj?.amount !== undefined,
    hasCurrency: obj?.currency !== undefined,
    moneyConstructor: Money?.toString()
  });

  if (!obj) {
    debugLog(`Invalid Money (${context})`, 'null or undefined');
    return false;
  }

  if (!(obj instanceof Money)) {
    // Try to coerce to Money if it has the right shape
    if (obj && typeof obj === 'object' && typeof obj.amount === 'number' && typeof obj.currency === 'string') {
      debugLog(`Attempting to coerce to Money (${context})`, obj);
      try {
        return new Money(obj.amount, obj.currency);
      } catch (e) {
        debugLog(`Coercion failed (${context})`, e.toString());
        return false;
      }
    }
    debugLog(`Invalid Money (${context})`, 'not a Money instance and cannot be coerced');
    return false;
  }

  return true;
};

const estimatedTotalPrice = (lineItems, marketplaceCurrency) => {
  // Log incoming data
  console.log('🔍 Debugging estimatedTotalPrice input:');
  console.log('Line items:', lineItems);
  console.log('Marketplace currency:', marketplaceCurrency);

  debugLog('estimatedTotalPrice input', {
    lineItems,
    marketplaceCurrency
  });

  const numericTotalPrice = lineItems.reduce((sum, lineItem) => {
    // Validate each lineItem's lineTotal
    console.log('🔍 Debugging lineItem.lineTotal:');
    console.log('Is Money instance:', lineItem.lineTotal instanceof Money);
    console.log('Object details:', lineItem.lineTotal);
    
    validateMoneyObject(lineItem.lineTotal, 'lineItem.lineTotal');
    const numericPrice = convertMoneyToNumber(lineItem.lineTotal);
    return new Decimal(numericPrice).add(sum);
  }, new Decimal(0));

  const currency =
    lineItems[0] && lineItems[0].unitPrice ? lineItems[0].unitPrice.currency : marketplaceCurrency;

  const result = new Money(
    convertUnitToSubUnit(numericTotalPrice.toNumber(), unitDivisor(currency)),
    currency
  );

  // Log result
  console.log('🔍 Debugging estimatedTotalPrice result:');
  console.log('Is Money instance:', result instanceof Money);
  console.log('Object details:', result);

  debugLog('estimatedTotalPrice result', result);
  return result;
};

const estimatedBooking = (bookingStart, bookingEnd, lineItemUnitType, timeZone = 'Etc/UTC') => {
  const duration = { start: bookingStart, end: bookingEnd };

  return {
    id: new UUID('estimated-booking'),
    type: 'booking',
    attributes: {
      ...duration,
    },
  };
};

// When we cannot speculatively initiate a transaction (i.e. logged
// out), we must estimate the transaction for booking breakdown. This function creates
// an estimated transaction object for that use case.
//
// We need to use the Template's backend to calculate the correct line items through thransactionLineItems
// endpoint so that they can be passed to this estimated transaction.
const estimatedCustomerTransaction = (
  lineItems,
  bookingStart,
  bookingEnd,
  lineItemUnitType,
  timeZone,
  process,
  processName,
  marketplaceCurrency
) => {
  const transitions = process?.transitions;
  const now = new Date();
  
  // Debug includeFor filtering
  console.log('[breakdown] Line items before filtering:', lineItems.map(li => ({
    code: li.code,
    includeFor: li.includeFor,
    hasCustomer: li.includeFor?.includes('customer'),
    hasProvider: li.includeFor?.includes('provider')
  })));
  
  const customerLineItems = lineItems.filter(item => item.includeFor.includes('customer'));
  const providerLineItems = lineItems.filter(item => item.includeFor.includes('provider'));
  
  console.log('[breakdown] After filtering:', {
    customerLineItems: customerLineItems.length,
    providerLineItems: providerLineItems.length,
    customerCodes: customerLineItems.map(li => li.code),
    providerCodes: providerLineItems.map(li => li.code)
  });
  
  const payinTotal = estimatedTotalPrice(customerLineItems, marketplaceCurrency);
  const payoutTotal = estimatedTotalPrice(providerLineItems, marketplaceCurrency);

  const bookingMaybe =
    bookingStart && bookingEnd
      ? { booking: estimatedBooking(bookingStart, bookingEnd, lineItemUnitType, timeZone) }
      : {};

  return {
    id: new UUID('estimated-transaction'),
    type: 'transaction',
    attributes: {
      createdAt: now,
      processName,
      lastTransitionedAt: now,
      lastTransition: transitions.REQUEST_PAYMENT,
      payinTotal,
      payoutTotal,
      lineItems: customerLineItems,
      transitions: [
        {
          createdAt: now,
          by: TX_TRANSITION_ACTOR_CUSTOMER,
          transition: transitions.REQUEST_PAYMENT,
        },
      ],
    },
    ...bookingMaybe,
  };
};

// Add robust Money validation and coercion helpers
const isMoney = obj => {
  return (
    obj &&
    typeof obj === 'object' &&
    typeof obj.amount === 'number' &&
    !isNaN(obj.amount) &&
    typeof obj.currency === 'string' &&
    obj.currency.length > 0
  );
};

const ensureMoney = (value, currency, fallbackAmount = 0) => {
  if (value instanceof Money) {
    return value;
  }
  if (isMoney(value)) {
    return new Money(Math.round(value.amount), value.currency);
  }
  if (typeof value === 'number' && !isNaN(value)) {
    return new Money(Math.round(value), currency);
  }
  console.error('Failed to coerce value to Money:', { value, currency });
  return new Money(fallbackAmount, currency);
};

const formatMoneySafely = (intl, money, fallbackText = 'Price unavailable') => {
  try {
    console.log('🔍 Debugging price formatting in EstimatedCustomerBreakdownMaybe:');
    console.log('Is Money instance:', money instanceof Money);
    console.log('Object details:', money);
    console.log('Money constructor:', Money);

    debugLog('Formatting money', {
      input: money,
      isMoneyInstance: money instanceof Money,
      amount: money?.amount,
      currency: money?.currency
    });

    // If it's not a Money instance but has the right shape, try to convert it
    if (!(money instanceof Money) && money?.amount !== undefined && money?.currency) {
      console.log('Attempting to convert to Money instance:', money);
      money = new Money(money.amount, money.currency);
    }

    if (!(money instanceof Money)) {
      throw new Error('Not a valid Money instance');
    }

    const formatted = formatMoney(intl, money);
    console.log('Money formatted successfully:', { input: money, output: formatted });
    debugLog('Money formatted successfully', { input: money, output: formatted });
    return formatted;
  } catch (e) {
    console.error('Money formatting failed:', e);
    debugLog('Money formatting failed', { error: e.toString(), input: money });
    return fallbackText;
  }
};

const EstimatedCustomerBreakdownMaybe = props => {
  const intl = useIntl();
  useEffect(() => {
    validateSDK();
    // debugLog('EstimatedCustomerBreakdownMaybe mounted', {
    //   props,
    //   sdkTypes,
    //   Money: sdkTypes.Money
    // });
  }, []);

  // Temporary guard to cope with accidental stringification
  const parseMaybe = v => {
    if (Array.isArray(v) || (v && typeof v === 'object')) return v;
    if (typeof v === 'string') {
      try { return JSON.parse(v); } catch {}
      try { return transit.read(v, { typeHandlers }); } catch {}
    }
    return null;
  };

  try {
    const { breakdownData: rawBreakdownData = {}, lineItems: rawLineItems, timeZone, currency, marketplaceName, processName } = props;
    
    // Apply parsing guard
    const breakdownData = parseMaybe(rawBreakdownData) || {};
    const lineItems = parseMaybe(rawLineItems) || [];
    
    console.log('[breakdown] Parsing guard applied:', {
      rawBreakdownDataType: typeof rawBreakdownData,
      rawLineItemsType: typeof rawLineItems,
      parsedBreakdownDataType: typeof breakdownData,
      parsedLineItemsType: typeof lineItems,
      isLineItemsArray: Array.isArray(lineItems)
    });
    const { startDate, endDate } = breakdownData;

    console.log('[breakdown] props:', { lineItems, currency, sdkTypes, breakdownData });

    // Debug breakdownData dates
    console.log('[breakdown] breakdownData dates:', {
      startDate: breakdownData?.startDate,
      endDate: breakdownData?.endDate,
      hasStartDate: !!breakdownData?.startDate,
      hasEndDate: !!breakdownData?.endDate,
      startDateType: typeof breakdownData?.startDate,
      endDateType: typeof breakdownData?.endDate
    });

    // Defensive guard: ensure lineItems is always an array
    const parsedLineItems = ensureArray(lineItems);

    // Fast visual of the lines
    console.table((parsedLineItems || []).map(li => ({
      code: li.code,
      amount: li.unitPrice?.amount,
      qty: li.quantity,
      includeFor: (li.includeFor || []).join(','),
      reversal: li.reversal
    })));

    debugLog('Processing props', { lineItems, parsedLineItems, currency });

    if (!parsedLineItems?.length) {
      console.log('[breakdown] No line items found');
      debugLog('No line items', null);
      return null;
    }

    // Debug LISTING_UNIT_TYPES
    console.log('[breakdown] LISTING_UNIT_TYPES:', LISTING_UNIT_TYPES);
    
    // Find the unit line item and validate it
    const unitLineItem = parsedLineItems.find(
      item => LISTING_UNIT_TYPES.includes(item.code) && !item.reversal
    );
    
    // Debug all line items and their codes
    console.log('[breakdown] All line item codes:', parsedLineItems.map(li => ({
      code: li.code,
      isInListingUnitTypes: LISTING_UNIT_TYPES.includes(li.code),
      isReversal: li.reversal,
      wouldMatch: LISTING_UNIT_TYPES.includes(li.code) && !li.reversal
    })));
    
    // Debug unit line item
    const unit = (Array.isArray(lineItems) ? lineItems : []).find(li => li.code === 'line-item/day');
    console.log('[breakdown] unit line:', unit);
    console.log('[breakdown] unitLineItem found:', unitLineItem);
    
    if (!unitLineItem) {
      console.log('[breakdown] No unit line item found - checking why...');
      console.log('[breakdown] Available codes:', parsedLineItems.map(li => li.code));
      console.log('[breakdown] Expected codes:', LISTING_UNIT_TYPES);
      debugLog('No unit line item found', null);
      return null;
    }

    debugLog('Unit line item', {
      code: unitLineItem.code,
      unitPrice: unitLineItem.unitPrice,
      isValid: validateMoneyObject(unitLineItem.unitPrice, 'unitLineItem.unitPrice')
    });

    const lineItemUnitType = unitLineItem.code;
    const numberOfDays = startDate && endDate ? differenceInCalendarDays(endDate, startDate) : 0;

    // Debug date calculation
    console.log('[breakdown] Date calculation:', {
      startDate,
      endDate,
      numberOfDays,
      startDateValid: !!startDate,
      endDateValid: !!endDate,
      bothDatesValid: !!(startDate && endDate)
    });

    // Early return if invalid booking duration
    if (numberOfDays <= 0) {
      console.log('[breakdown] Invalid booking duration - returning null');
      console.log('[breakdown] numberOfDays:', numberOfDays, 'startDate:', startDate, 'endDate:', endDate);
      debugLog('Invalid booking duration', { startDate, endDate, numberOfDays });
      return null;
    }

    // Remove all frontend discount logic. Only use the provided lineItems.
    let adjustedLineItems = parsedLineItems;

    let process = null;
    try {
      process = getProcess(processName);
    } catch (e) {
      return (
        <div className={css.error}>
          <FormattedMessage id="OrderPanel.unknownTransactionProcess" />
        </div>
      );
    }

    const shouldHaveBooking = [LINE_ITEM_DAY, LINE_ITEM_NIGHT].includes(lineItemUnitType);
    const hasLineItems = parsedLineItems && parsedLineItems.length > 0;
    const hasRequiredBookingData = !shouldHaveBooking || (startDate && endDate);

    const tx =
      hasLineItems && hasRequiredBookingData
        ? estimatedCustomerTransaction(
            adjustedLineItems,
            startDate,
            endDate,
            lineItemUnitType,
            timeZone,
            process,
            processName,
            currency
          )
        : null;

    return tx ? (
      <OrderBreakdown
        className={css.receipt}
        userRole="customer"
        transaction={tx}
        booking={tx.booking}
        timeZone={timeZone}
        currency={currency}
        marketplaceName={marketplaceName}
      />
    ) : null;
  } catch (error) {
    debugLog('Component Error', {
      error: error?.toString(),
      stack: error?.stack
    });
    return null;
  }
};

export default EstimatedCustomerBreakdownMaybe;
