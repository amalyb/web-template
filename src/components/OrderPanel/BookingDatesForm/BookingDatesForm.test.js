import React from 'react';
import { shallow } from 'enzyme';
import Decimal from 'decimal.js';

import { types as sdkTypes } from '../../../util/sdkLoader';
import { renderShallow, renderDeep } from '../../../util/test-helpers';
import { fakeIntl } from '../../../util/test-data';
import { LINE_ITEM_NIGHT, LISTING_UNIT_TYPES } from '../../../util/types';
import { timeOfDayFromTimeZoneToLocal } from '../../../util/dates';

import { OrderBreakdown } from '../../../components';

import { BookingDatesFormComponent } from './BookingDatesForm';
import EstimatedCustomerBreakdownMaybe from '../EstimatedCustomerBreakdownMaybe';

const { Money } = sdkTypes;

const noop = () => null;
const lineItems = [
  {
    code: 'line-item/night',
    unitPrice: new Money(1099, 'USD'),
    quantity: new Decimal(2),
    includeFor: ['customer', 'provider'],
    lineTotal: new Money(2198, 'USD'),
    reversal: false,
  },
];

describe('BookingDatesForm', () => {
  it('matches snapshot without selected dates', () => {
    const tree = renderShallow(
      <BookingDatesFormComponent
        lineItemUnitType={LINE_ITEM_NIGHT}
        intl={fakeIntl}
        dispatch={noop}
        onSubmit={v => v}
        price={new Money(1099, 'USD')}
        bookingDates={{}}
        startDatePlaceholder="today"
        endDatePlaceholder="tomorrow"
        fetchLineItemsInProgress={false}
        onFetchTransactionLineItems={noop}
        onFetchTimeSlots={noop}
        lineItems={lineItems}
        currency="USD"
        marketplaceName="MarketplaceX"
        processName="default-booking"
        dayCountAvailableForBooking={90}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});

describe('EstimatedCustomerBreakdownMaybe', () => {
  it('renders nothing if nightly is missing start and end date', () => {
    expect(
      renderDeep(
        <EstimatedCustomerBreakdownMaybe
          lineItems={lineItems}
          currency="USD"
          marketplaceName="MarketplaceX"
          processName="default-booking"
        />
      )
    ).toBeFalsy();
  });
  it('renders nothing if nightly is missing end date', () => {
    const data = {
      startDate: new Date(),
    };
    expect(
      renderDeep(
        <EstimatedCustomerBreakdownMaybe
          breakdownData={data}
          lineItems={lineItems}
          currency="USD"
          marketplaceName="MarketplaceX"
          processName="default-booking"
        />
      )
    ).toBeFalsy();
  });
  it('renders breakdown with correct transaction data', () => {
    const startDate = new Date(2017, 3, 14, 0, 0, 0);
    const endDate = new Date(2017, 3, 16, 0, 0, 0);
    const props = {
      breakdownData: {
        startDate,
        endDate,
      },
      lineItems,
      currency: 'USD',
      marketplaceName: 'MarketplaceX',
      processName: 'default-booking',
    };

    const tree = shallow(<EstimatedCustomerBreakdownMaybe {...props} />);
    const breakdown = tree.find(OrderBreakdown);
    const { userRole, transaction, booking } = breakdown.props();
    const unitLineItem = transaction.attributes.lineItems.find(
      item => LISTING_UNIT_TYPES.includes(item.code) && !item.reversal
    );

    expect(unitLineItem.code).toEqual(LINE_ITEM_NIGHT);
    expect(userRole).toEqual('customer');

    // booking data doesn't get converted inside EstimatedCustomerBreakdownMaybe
    expect(booking.attributes.start).toEqual(startDate);
    expect(booking.attributes.end).toEqual(endDate);

    expect(transaction.attributes.payinTotal).toEqual(new Money(2198, 'USD'));
    expect(transaction.attributes.payoutTotal).toEqual(new Money(2198, 'USD'));
    expect(transaction.attributes.lineItems).toEqual([
      {
        code: 'line-item/night',
        unitPrice: new Money(1099, 'USD'),
        quantity: new Decimal(2),
        includeFor: ['customer', 'provider'],
        lineTotal: new Money(2198, 'USD'),
        reversal: false,
      },
    ]);
  });
});
