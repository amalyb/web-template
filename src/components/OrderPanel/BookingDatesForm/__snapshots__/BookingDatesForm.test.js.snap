// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingDatesForm matches snapshot without selected dates 1`] = `
<ReactFinalForm
  bookingDates={Object {}}
  dispatch={[Function]}
  endDatePlaceholder="tomorrow"
  fetchLineItemsError={null}
  intl={
    Object {
      "formatDate": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelativeTime": [Function],
      "formatTime": [Function],
      "now": [Function],
    }
  }
  lineItemUnitType="line-item/night"
  lineItems={
    Array [
      Object {
        "code": "line-item/night",
        "includeFor": Array [
          "customer",
          "provider",
        ],
        "lineTotal": Money {
          "_sdkType": "Money",
          "amount": 2198,
          "currency": "USD",
        },
        "quantity": "2",
        "reversal": false,
        "unitPrice": Money {
          "_sdkType": "Money",
          "amount": 1099,
          "currency": "USD",
        },
      },
    ]
  }
  monthlyTimeSlots={null}
  onFetchTimeSlots={[Function]}
  onSubmit={[Function]}
  render={[Function]}
  startDatePlaceholder="today"
  unitPrice={
    Money {
      "_sdkType": "Money",
      "amount": 1099,
      "currency": "USD",
    }
  }
/>
`;
